integration/utils/follower.go

```diff
@@ -0,0 +1,231 @@
+package utils
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	flowsdk "github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/client"
+	"go.uber.org/atomic"
+
+	"github.com/rs/zerolog"
+)
+
+type TxFollower interface {
+	// CompleteChanByID returns a channel that is closed when the transaction is complete.
+	CompleteChanByID(ID flowsdk.Identifier) <-chan struct{}
+
+	// Height returns the last acted upon block height.
+	Height() uint64
+	// BlockID returns the last acted upon block ID.
+	BlockID() flowsdk.Identifier
+
+	// Stop sends termination signal to the follower.
+	// It waits until the follower is stopped.
+	Stop()
+}
+
+type followerOption func(f *txFollowerImpl)
+
+func WithBlockHeight(height uint64) followerOption {
+	return func(f *txFollowerImpl) { f.height = height }
+}
+
+func WithLogger(logger zerolog.Logger) followerOption {
+	return func(f *txFollowerImpl) { f.logger = logger }
+}
+
+func WithInteval(interval time.Duration) followerOption {
+	return func(f *txFollowerImpl) { f.interval = interval }
+}
+
+type txFollowerImpl struct {
+	logger zerolog.Logger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	client *client.Client
+
+	interval time.Duration
+
+	inprogress *atomic.Int64
+
+	mu      *sync.RWMutex
+	height  uint64
+	blockID flowsdk.Identifier
+
+	txToChan sync.Map
+
+	stopped chan struct{}
+}
+
+type txInfo struct {
+	submisionTime time.Time
+
+	C chan struct{}
+}
+
+// NewTxFollower creates a new follower that tracks the current block height
+// and can notify on transaction completion.
+func NewTxFollower(ctx context.Context, client *client.Client, opts ...followerOption) (TxFollower, error) {
+	newCtx, cancel := context.WithCancel(ctx)
+
+	f := &txFollowerImpl{
+		client: client,
+		ctx:    newCtx,
+		cancel: cancel,
+		logger: zerolog.Nop(),
+		mu:     &sync.RWMutex{},
+
+		inprogress: atomic.NewInt64(0),
+
+		stopped:  make(chan struct{}),
+		interval: 100 * time.Millisecond,
+	}
+
+	for _, opt := range opts {
+		opt(f)
+	}
+
+	if f.height == 0 {
+		hdr, err := client.GetLatestBlockHeader(newCtx, true)
+		if err != nil {
+			return nil, err
+		}
+		f.height = hdr.Height
+		f.blockID = hdr.ID
+	}
+
+	go f.follow()
+
+	return f, nil
+}
+
+func (f *txFollowerImpl) follow() {
+	t := time.NewTicker(f.interval)
+	defer t.Stop()
+	defer close(f.stopped)
+
+Loop:
+	for lastBlockTime := time.Now(); ; <-t.C {
+		select {
+		case <-f.ctx.Done():
+			return
+		default:
+		}
+
+		blockResolutionStart := time.Now()
+		block, err := f.client.GetBlockByHeight(f.ctx, f.height+1)
+		if err != nil {
+			continue
+		}
+
+		var totalTxs uint64
+		for _, guaranteed := range block.CollectionGuarantees[:] {
+			col, err := f.client.GetCollection(f.ctx, guaranteed.CollectionID)
+			if err != nil {
+				continue Loop
+			}
+			for _, tx := range col.TransactionIDs {
+				totalTxs++
+				if ch, loaded := f.txToChan.LoadAndDelete(tx.Hex()); loaded {
```

Comment:
Suggestion: Having a counter for TX IDs discovered in blocks which are unknown? Why? Let's say multiple loaders are used concurrently, or a loader and additional ways of sending TXs... it would be nice to account for 'unknown' TXs in the log line below, or?